### PR TITLE
refactor(streaming): use a separate degree table for hash join

### PR DESCRIPTION
**Pull request by yuhao-su**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Previously, the update degree operation in hash join will update the whole row, which is expensive for storage.

- use a separate degree table for hash join
- do a lot of rename to make the name of keys clear.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
#3254

### CODE REVIEW

File: src/stream/src/executor/managed_state/join/mod.rs [hunk 1]
```diff
@@ -141,6 +141,24 @@ impl EncodedJoinRow {
...
+        row_data_types: &[DataType],
+        state_order_key_indices: &[usize],
+    ) -> StreamExecutorResult<Row> {
+        let degree_datum = Some(ScalarImpl::Int64(self.degree as i64));
+        let prefix = self
+            .decode_row(row_data_types)?
+            .by_indices(state_order_key_indices);
+        let schemaed_degree = state_table::append_pk_prefix(Row::new(vec![degree_datum]), prefix);
+        Ok(schemaed_degree)
+    }
```
Comment (yuhao-su) [inline src/stream/src/executor/managed_state/join/mod.rs:161]: The degree table schema is pk of table + degree. To fit the state table API, I have to manually add the pk to the prefix of degree. We may need a kv API for state table

File: src/stream/src/executor/managed_state/join/mod.rs [hunk 2]
```diff
@@ -85,24 +88,15 @@ impl JoinRow {
...
-        }
+    /// Return row and degree in `Row` format. The degree part will be inserted in degree table
+    /// later, so a pk prefix will be added.
+    ///
+    /// * `state_order_key_indices` - the order key of `row`
+    pub fn into_table_rows(self, state_order_key_indices: &[usize]) -> (Row, Row) {
+        let degree_datum = Some(ScalarImpl::Int64(self.degree as i64));
+        let pk_prefix = self.row.by_indices(state_order_key_indices);
+        let degree = state_table::append_pk_prefix(Row::new(vec![degree_datum]), pk_prefix);
+        (self.row, degree)
```
Comment (yuhao-su) [inline src/stream/src/executor/managed_state/join/mod.rs:99]: I'm creating a schemaed degree by taking the order key of the row as prefix. We may optimize this.

File: src/storage/src/table/storage_table/iter_utils.rs [hunk 3]
```diff
@@ -81,6 +84,71 @@ where
...
 }
 
+/// Merge two streams of primary key and rows into a single stream, sorted by primary key.
+/// We should ensure that the primary key from different streams are unique.
+#[try_stream(ok = (Cow<'a, Row>, Cow<'a, Row>), error = StorageError)]
+pub async fn merge_by_pk<'a, S>(
+    stream1: S,
+    pk_indices1: &'a [usize],
+    stream2: S,
+    pk_indices2: &'a [usize],
```
Comment (yuhao-su) [inline src/storage/src/table/storage_table/iter_utils.rs:94]: The lifetime annotation for pk_indices seems unnecessary. But removing those won't compile.

File: src/frontend/src/optimizer/plan_node/stream_hash_join.rs [hunk 4]
```diff
@@ -228,34 +230,59 @@ impl ToStreamProst for StreamHashJoin {
...
+    // key when there are duplicate value in join key indices.
+    let degree_table_dist_keys = internal_table_dist_keys
+        .iter()
+        .map(|idx| {
+            join_key_indices
+                .iter()
+                .position(|v| v == idx)
+                .expect("join key should contain dist key.")
+        })
+        .collect_vec();
```
Comment (yuhao-su) [inline src/frontend/src/optimizer/plan_node/stream_hash_join.rs:256]: PLTA @st1page

**Comment by st1page** [comment]
reviewing... I will continue reviewing it tomorrow.

**Review by wcy-fdu — COMMENTED** [review]
Could you please do some benchmark on original `hash_join` and your new design?

File: src/common/src/array/data_chunk_iter.rs [hunk 5]
```diff
@@ -249,6 +249,19 @@ impl Row {
         &EMPTY_ROW
     }
 
+    /// Compare two rows' stream key
+    pub fn eq_by_pk(
+        row1: impl AsRef<Self>,
+        pk_indices1: &[usize],
+        row2: impl AsRef<Self>,
+        pk_indices2: &[usize],
+    ) -> bool {
```
Comment (st1page) [inline src/common/src/array/data_chunk_iter.rs:258]:
```suggestion
    pub fn eq_by_keys(
        row1: impl AsRef<Self>,
        key1: &[usize],
        row2: impl AsRef<Self>,
        key2: &[usize],
    ) -> bool {
```

File: src/frontend/src/catalog/table_catalog.rs [hunk 6]
```diff
@@ -138,7 +138,7 @@ impl TableCatalog {
         self.distribution_key.as_ref()
     }
 
-    pub fn to_state_table_prost(&self) -> ProstTable {
+    pub fn to_internal_table_prost(&self) -> ProstTable {
```
Comment (st1page) [inline src/frontend/src/catalog/table_catalog.rs:141]: why we need change this name?

**Review by st1page — COMMENTED** [review]
Generally LGTM, confused with the rename of `to_state_table_prost`

**Review by st1page — APPROVED** [review]
LGTM

File: src/frontend/src/catalog/table_catalog.rs (hunk 6 above)
Comment (yuhao-su) [reply src/frontend/src/catalog/table_catalog.rs:141]: currently, only internal tables need to fill in a placeholder, while mv does not need this.

File: src/common/src/array/data_chunk_iter.rs (hunk 5 above)
Comment (yuhao-su) [reply src/common/src/array/data_chunk_iter.rs:258]: Actually, this function requires that the keys are ordered. In this case, would it be better named ordered_key?

**Comment by yuhao-su** [comment]
After some bench, I noticed that the source throughput drop to 2/3 of main branch. My guess is it is mainly because we need to issue more (twice) the read operation after we separate the table. We still need further investigation to confirm.
image
image

**Pull request closed without merging**